DCOS-46615 - Retry downloading pkgpanda pkg if file size incorrect

mrnugget opened the pull request:
## High-level description

AWS S3, from which we download pre-built pkgpanda packages, might sometimes end the connection before the package has been fully downloaded. (See https://github.com/aws/aws-sdk-js/issues/312 and https://github.com/boto/boto3/issues/540 for more context on this)

In our case that leads to invalid `tar.xz` archives which cause `unexpected end of input` errors and result in faulty builds of DC/OS (`dcos_generate_config.sh`) which contain these invalid archives and then result in install-time errors for users. See this comment for an investigation into how this can affect installation errors: https://jira.mesosphere.com/browse/DCOS_OSS-4097?focusedCommentId=219259&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-219259

This change here adds a sanity check to the `download` function used by pkgpanda. It checks whether the final file size (after flushing the file to disk) matches the `Content-Length` we get from S3. If it doesn't match, it sleep for 2, 4, 8, ... seconds and then retries until it either downloads the file completely or the retries are exhausted. In that case, it raises an exception.

## Corresponding DC/OS tickets (obligatory)

These DC/OS JIRA ticket(s) must be updated (ideally closed) in the moment this PR lands:

  - [[DCOS-46615] Pkgpanda - package dcos-image - xz: (stdin): Unexpected end of input - JIRA](https://jira.mesosphere.com/browse/DCOS-46615)

## Related tickets (optional)

Other tickets related to this change:

  - [[DCOS_OSS-4097] pkgpanda: Explicitly create user groups for all users - JIRA](https://jira.mesosphere.com/browse/DCOS_OSS-4097)


## Checklist for all PRs

  - [x] Added a comprehensible changelog entry to `CHANGES.md` or explain why this is not a user-facing change: it's internal behaviour that's not exposed to the user
  - [x] Included a test which will fail if code is reverted but test is not. If there is no test please explain here: I didn't include a test yet, since that would need to involve setting up a mock http server etc. and I wanted to get input on this first
  - [x] Read the [DC/OS contributing guidelines](https://github.com/dcos/dcos/blob/master/contributing.md)
  - [x] Followed relevant code rules [Rules for Packages and Systemd](https://github.com/dcos/dcos/tree/master/docs)


## Checklist for component/package updates:

If you are changing components or packages in DC/OS (e.g. you are bumping the sha or ref of anything underneath `packages`), then in addition to the above please also include:

  - [ ] Change log from the last version integrated (this should be a link to commits for easy verification and review): [example](https://github.com/dcos/dcos-mesos-modules/compare/f6fa27d7c40f4207ba3bb2274e2cfe79b62a395a...6660b90fbbf69a15ef46d0184e36755881d6a5ae)
  - [ ] Test Results: [link to CI job test results for component]
  - [ ] Code Coverage (if available): [link to code coverage report]
___
**PLEASE FILL IN THE TEMPLATE ABOVE** / **DO NOT REMOVE ANY SECTIONS ABOVE THIS LINE**


## Instructions and review process

**What is the review process and when will my changes land?**

All PRs require 2 approvals using GitHub's [pull request reviews](https://help.github.com/articles/about-pull-request-reviews/).

Reviewers should be:
* Developers who understand the code being modified.
* Developers responsible for code that interacts with or depends on the code being modified.

It is best to proactively ask for 2 reviews by @mentioning the candidate reviewers in the PR comments area. The responsibility is on the developer submitting the PR to follow-up with reviewers and make sure a PR is reviewed in a timely manner. Once a PR has **2 ship-it's**, **no red reviews**, and **all tests are green** it will be included in the [next train](https://github.com/dcos/dcos/blob/master/contributing.md).

gpaul commented on pkgpanda/util.py:
```diff
@@ -150,6 +151,38 @@ def get_requests_retry_session(max_retries=4, backoff_factor=1, status_forcelist
+            pass
+
+    while True:
+        with open(out_filename, "w+b") as f:
```
I think you should drop the `+` here since you're not reading, only writing.

mrnugget replied:
Also taken from the existing code, but it's probably safe to drop it.

gpaul commented on pkgpanda/util.py:
```diff
@@ -150,6 +151,38 @@ def get_requests_retry_session(max_retries=4, backoff_factor=1, status_forcelist
+
+    while True:
+        with open(out_filename, "w+b") as f:
+            r = get_requests_retry_session().get(url, stream=True)
```
By default, `get` will follow redirects (http://docs.python-requests.org/en/v3.0.0/_modules/requests/api/#get). If you really want to ensure that the request did not get redirected you should pass `allow_redirects=False` here.

gpaul commented on pkgpanda/util.py:
```diff
@@ -150,6 +151,38 @@ def get_requests_retry_session(max_retries=4, backoff_factor=1, status_forcelist
+    while True:
+        with open(out_filename, "w+b") as f:
+            r = get_requests_retry_session().get(url, stream=True)
+            if r.status_code == 301:
```
Why do you care about redirects? It seems reasonable that the CDN / S3 would redirect you to a specific download URL.

mrnugget replied:
As I wrote in Slack: I — personally — don't care 😄 I've taken this from the existing (untested) code that was already there, presumably for years. And when working with legacy code that doesn't have tests to confirm something. I tread lightly and only really change what needs to be changed. Or, in other words: if I don't know why code is there and don't know why it _shouldn't_ be there I'll leave it alone — for now, because I don't want to break something that I didn't know existed.

gpaul replied:
SGTM

gpaul commented on pkgpanda/util.py:
```diff
@@ -150,6 +151,38 @@ def get_requests_retry_session(max_retries=4, backoff_factor=1, status_forcelist
+
+            for chunk in r.iter_content(chunk_size=4096):
+                f.write(chunk)
+            f.flush()
```
Consider maintaining "total bytes read" by accumulating `len(chunk)` instead of `stat`ing the file and explicitly flushing it.

mrnugget replied:
That's a good idea!

gpaul commented on pkgpanda/util.py:
```diff
@@ -150,6 +151,38 @@ def get_requests_retry_session(max_retries=4, backoff_factor=1, status_forcelist
+            if final_size == content_length:
+                return r
+            else:
+                f.close()
```
Calling `f.close()` is often a sign that you could restructure your code, as the context manager `with open(...) as f:` will do that automatically when you leave its scope.

mrnugget replied:
Ah, yes, I wanted to clean this up after confirming general approach. Let's see if we can get rid of a lot of it when using `retrying`...

gpaul commented on pkgpanda/util.py:
```diff
@@ -150,6 +151,38 @@ def get_requests_retry_session(max_retries=4, backoff_factor=1, status_forcelist
+            else:
+                f.close()
+                cleanup()
+                retries -= 1
```
There's a very lovely little python library called `retrying` (https://pypi.org/project/retrying/) that let's you annotate a function (even an inline function) with retry logic of this sort.) We should use that instead of rolling our own.

For example:
https://github.com/dcos/dcos/blob/5391993fa0ca8457dae4d2126d2c0c515dd8264a/packages/dcos-integration-test/extra/util/delete_ec2_volume.py#L46
https://github.com/dcos/dcos/blob/57350f12eb7c1ed82eeb7e8944c72776b936160d/packages/dcos-integration-test/extra/test_networking.py#L60-L62

I like declaring little inner functions in my larger function and annotating the inner ones with retrying logic.

adamtheturtle replied:
I see we had the same thought! https://github.com/dcos/dcos/pull/4549/files#r258438916

mrnugget replied:
Switched to `retrying` — it's lovely!

gpaul commented on pkgpanda/util.py:
```diff
@@ -167,14 +200,7 @@ def download(out_filename, url, work_dir, rm_on_error=True):
-                r.raise_for_status()
-                for chunk in r.iter_content(chunk_size=4096):
-                    f.write(chunk)
+            stream_remote_file_with_retries(out_filename, url)
```
I think `_with_retries` is an implementation detail, let's just call the function "_download_file" ?

mrnugget replied:
I wanted to differentiate it from the existing `download` function and make its possibly slow behaviour explicit, but I get your point.

I'll rename it to `_download_remote_file`, since `download` has two branches: "download" local file or download remote file. The latter is what's encapsulated in this function.

adamtheturtle commented on pkgpanda/util.py:
```diff
@@ -150,6 +151,38 @@ def get_requests_retry_session(max_retries=4, backoff_factor=1, status_forcelist
     return session
 
 
+def stream_remote_file_with_retries(out_filename, url, retries=4):
```
Can we use a known-good retrying library. We already have one on clusters (see [here](https://github.com/dcos/dcos/blob/86a6934657c3d1c22f77d3cd21e112f940988043/ssh/test_integration.py#L13) for an example use).

adamtheturtle commented on pkgpanda/util.py:
```diff
@@ -150,6 +151,38 @@ def get_requests_retry_session(max_retries=4, backoff_factor=1, status_forcelist
     return session
 
 
+def stream_remote_file_with_retries(out_filename, url, retries=4):
```
`retries=4`! Let's explain all of these choices, even if the explanation is "this is my best guess for a compromise and it seems to work". That means that the next maintainer will know whether it is safe to change.

mrnugget replied:
This is taken from here:

https://github.com/dcos/dcos/blob/2897e2459e460033fb2b72099d4aac29e2b58937/pkgpanda/util.py#L138

No idea why it's `4` but I thought I'd keep the behaviour the same.

adamtheturtle commented on pkgpanda/util.py:
```diff
@@ -150,6 +151,38 @@ def get_requests_retry_session(max_retries=4, backoff_factor=1, status_forcelist
+    def cleanup():
+        try:
+            os.remove(out_filename)
+        except Exception:
```
Can you perhaps be more particular with the exceptions that you wish to ignore?

mrnugget replied:
This is also taken from here: https://github.com/dcos/dcos/blob/2897e2459e460033fb2b72099d4aac29e2b58937/pkgpanda/util.py#L185-L189

adamtheturtle commented on pkgpanda/util.py:
```diff
@@ -150,6 +151,38 @@ def get_requests_retry_session(max_retries=4, backoff_factor=1, status_forcelist
+        with open(out_filename, "w+b") as f:
+            r = get_requests_retry_session().get(url, stream=True)
+            if r.status_code == 301:
+                raise Exception("got a 301")
```
Where do you expect this error message to be seen - what do you expect someone who sees it to do?

mrnugget replied:
See here: https://github.com/dcos/dcos/pull/4549#discussion_r259736459

adamtheturtle commented on pkgpanda/util.py:
```diff
@@ -150,6 +151,38 @@ def get_requests_retry_session(max_retries=4, backoff_factor=1, status_forcelist
+            final_size = os.fstat(f.fileno()).st_size
+            if final_size == content_length:
+                return r
+            else:
```
There is no need for `else` here.

mrnugget replied:
Fixed that :)

mrnugget commented on pkgpanda/test_util.py:
```diff
@@ -412,3 +416,53 @@ def test_write_string(tmpdir):
+
+class MockDownloadServerRequestHandler(BaseHTTPRequestHandler):
+    def do_GET(self):  # noqa: N802
+        self.send_response(requests.codes.ok)
```
Will do that, yes!

And, yes, `BaseHTTPRequestHandler` calls `flush`: https://github.com/python/cpython/blob/ca5e6aaea5168bff274f3e98d7970bfa1bc366a2/Lib/http/server.py#L415

mrnugget commented on pkgpanda/test_util.py:
```diff
@@ -412,3 +416,53 @@ def test_write_string(tmpdir):
+def test_stream_remote_file_with_retries(tmpdir):
+    def _get_free_port():
+        s = socket.socket(socket.AF_INET, type=socket.SOCK_STREAM)
+        s.bind(('localhost', 0))
```
Thanks! The `_get_free_port` was a straight prototyping-let's-get-something-to-run-copy from an example. Now I learned something and will use your suggestion :)

mrnugget commented on pkgpanda/test_util.py:
```diff
@@ -412,3 +416,53 @@ def test_write_string(tmpdir):
+    assert mock_server.requests_received == 2
+
+    with open(out_file) as f:
+        assert f.read() == 'foobar'
```
Your attention to consistency is balm for this nitpicking programmer's soul! :)

mrnugget commented:
@mesosphere-mergebot label Ready for Review

mrnugget commented:
@gpaul @adamtheturtle Can you re-review? I removed the useless legacy code, worked in all of your suggestions and switched to `retrying` – the only thing left is the long-running thread for the mock http server, which will run until the end of the test suite, I guess. I don't think that's a huge issue, but I'll look into it.

mrnugget commented:
@mesosphere-mergebot bump-ee

gpaul commented:
@mesosphere-mergebot sync

gpaul commented:
@mesosphere-mergebot sync
